src/DynamoCoreWpf/ViewModels/Core/HomeWorkspaceViewModel.cs

```diff
@@ -91,10 +92,27 @@ public override void OnNodeModified(NodeModel obj)
 
         private void hwm_SetNodeDeltaState(object sender, DeltaComputeStateEventArgs e)
         {
-            var nodeGuids = e.NodeGuidList;
+            // Using Nodes here is not thread safe, as nodes can be added/removed by the UI thread midway.
+            // Dispatching this to the UI thread helps avoid concurrency issues and seems more correct.
+            if (!DynamoViewModel.Model.ShutdownRequested)
+            {
+                DynamoViewModel.UIDispatcher.Invoke(() => UpdateNodesDeltaState(e.NodeGuidList, e.GraphExecuted));
+            }
+            else
+            {
+                // However, when Dynamo is shutting down, a funny deadlock situation occurs.
+                // The main thread is waiting for the scheduler to finish executing.
+                // So in order to avoid it, we do the update on the scheduler thread.
+                // Concurrency should not be an issue now that the UI is blocked waiting for shutdown.
+                UpdateNodesDeltaState(e.NodeGuidList, e.GraphExecuted);
+            }
+        }
+
+        private void UpdateNodesDeltaState(List<Guid> nodeGuids, bool graphExecuted)
+        {
             // if runsettings is manual, and if the graph is not executed, then turing on showrunpreview 
             //should turn on showexectionpreview on every node.
-            if (nodeGuids.Count == 0 && !e.GraphExecuted)
+            if (nodeGuids.Count == 0 && graphExecuted)
```

Comment:
Are you sure this needs to be negated here? Was it a bug?

Reply:
Oops. That's my mistake actually. Didn't mean to change that. Good catch!